runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/ParDos.java

```diff
@@ -0,0 +1,348 @@
+    byte[] serializedFn =
+        fnSpec.getSpec().getParameter().unpack(BytesValue.class).getValue().toByteArray();
+    return (DoFnAndMainOutput)
+        SerializableUtils.deserializeFromByteArray(serializedFn, "Custom DoFn And Main Output tag");
+  }
+
+  private static Optional<RunnerApi.Parameter> toProto(Parameter parameter) {
+    Cases<Type> runnerApiCases =
+        new Cases<Type>() {
+          @Override
+          public Type dispatch(StartBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(FinishBundleContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(ProcessContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(OnTimerContextParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(WindowParameter p) {
+            return Type.WINDOW;
+          }
+
+          @Override
+          public Type dispatch(RestrictionTrackerParameter p) {
+            return Type.RESTRICTION_TRACKER;
+          }
+
+          @Override
+          public Type dispatch(StateParameter p) {
+            return null;
+          }
+
+          @Override
+          public Type dispatch(TimerParameter p) {
+            return null;
+          }
+        };
+    Type parameterType = parameter.match(runnerApiCases);
+    if (parameterType == null) {
+      return Optional.absent();
+    }
+    return Optional.of(RunnerApi.Parameter.newBuilder().setType(parameterType).build());
+  }
+
+  private static SideInput toProto(PCollectionView<?> view) {
+    Builder builder = SideInput.newBuilder();
+    builder.setAccessPattern(
+        FunctionSpec.newBuilder()
+            .setUrn(view.getViewFn().getMaterialization().getUrn())
+            .build());
+    builder.setViewFn(toProto(view.getViewFn()));
+    builder.setWindowMappingFn(toProto(view.getWindowMappingFn()));
+    return builder.build();
+  }
+
+  public static PCollectionView<?> fromProto(
+      String id, SideInput sideInput, RunnerApi.PTransform parDoTransform, Components components)
+      throws IOException {
+    TupleTag<?> tag = new TupleTag<>(id);
+    WindowMappingFn<?> windowMappingFn = windowMappingFnFromProto(sideInput.getWindowMappingFn());
+    ViewFn<?, ?> viewFn = viewFnFromProto(sideInput.getViewFn());
+
+    RunnerApi.PCollection inputCollection =
+        components.getPcollectionsOrThrow(parDoTransform.getInputsOrThrow(id));
+    WindowingStrategy<?, ?> windowingStrategy =
+        WindowingStrategies.fromProto(
+            components.getWindowingStrategiesOrThrow(inputCollection.getWindowingStrategyId()),
+            components);
+    Coder<?> elemCoder =
+        Coders.fromProto(components.getCodersOrThrow(inputCollection.getCoderId()), components);
+    Coder<Iterable<WindowedValue<?>>> coder =
```

Comment:
This should be keyed on the materialization. This is the appropriate coder for our current "iterable of full windowed values" materialization. It is OK to not implement any others, but there should be a `checkArgument` or a conditional with a failing else branch with a meaningful message in an `Unsupported` exception.